### PR TITLE
Update to (beta of) latest broccoli-typescript-compiler

### LINKED ISSUE #7: Cannot generate component
A couple issues here, some of which I'll see if I can open a PR for:

1. This is currently looking for `../../lib/<a bunch of things>`; it should switch to `ember-cli/lib/<a bunch of things>` instead so it's not depending on install location.
2. This is looking for `ember-cli/lib/utilities/string`; this doesn't appear to be a *thing* at all (I don't know when it was, but I assume it was at some point). It should look instead for `ember-cli-string-utilities`.
3. Even after making those fixes, I get this error:

    ```sh
    $ ember generate component my-foo
    installing component
    Unknown blueprint: component
    ```

### LINKED ISSUE #23: Unknown compiler option 'compileOnSave'
Running ```tsc``` directly, ```compileOnSave``` works, but ```ember serve``` command throws an error. 

Also the compiler seems to be running 3 times (using default tsconfig.json as recommended by the repo)

```
Livereload server on http://localhost:49153
Serving on http://localhost:4200/
Unknown compiler option 'compileOnSave'.
Unknown compiler option 'compileOnSave'.
Unknown compiler option 'compileOnSave'.
```

```
ember --version
ember-cli: 2.12.1
node: 7.4.0
os: linux x64
```
and
```
tsc -v
Version 2.2.2
```

package.json TS stuff

```
 "@types/ember": "^2.7.37",
 "@types/underscore": "^1.8.0",
 "ember-cli-typescript": "github:emberwatch/ember-cli-typescript",
"typescript": "^2.2.2"
```

### LINKED ISSUE #38: TypeError: Cannot read property 'length' of undefined
I do get this error fairly often:

```
$ ember s
Livereload server on http://localhost:49153
The Broccoli Plugin: [broccoli-typescript-compiler] failed with:
TypeError: Cannot read property 'length' of undefined
    at createSourceFile (/path/to/my/project/node_modules/typescript/lib/typescript.js:15396:109)
    at parseSourceFileWorker (/path/to/my/project/node_modules/typescript/lib/typescript.js:15328:26)
    at Object.parseSourceFile (/path/to/my/project/node_modules/typescript/lib/typescript.js:15277:26)
    at Object.createSourceFile (/path/to/my/project/node_modules/typescript/lib/typescript.js:15131:29)
    at SourceCache.getSourceFile (/path/to/my/project/node_modules/broccoli-typescript-compiler/dist/plugin.js:98:43)
    at Object.getSourceFile (/path/to/my/project/node_modules/broccoli-typescript-compiler/dist/plugin.js:300:85)
    at findSourceFile (/path/to/my/project/node_modules/typescript/lib/typescript.js:67716:29)
    at processSourceFile (/path/to/my/project/node_modules/typescript/lib/typescript.js:67647:27)
    at /path/to/my/project/node_modules/typescript/lib/typescript.js:67757:17
    at Object.forEach (/path/to/my/project/node_modules/typescript/lib/typescript.js:1443:30)
    at processReferencedFiles (/path/to/my/project/node_modules/typescript/lib/typescript.js:67755:16)
    at findSourceFile (/path/to/my/project/node_modules/typescript/lib/typescript.js:67740:21)
    at processSourceFile (/path/to/my/project/node_modules/typescript/lib/typescript.js:67647:27)
    at processRootFile (/path/to/my/project/node_modules/typescript/lib/typescript.js:67535:13)
    at Object.createProgram (/path/to/my/project/node_modules/typescript/lib/typescript.js:66845:21)
    at Compiler.createProgram (/path/to/my/project/node_modules/broccoli-typescript-compiler/dist/plugin.js:236:22)

The broccoli plugin was instantiated at:
    at TypeScript.Plugin (/path/to/my/project/node_modules/broccoli-plugin/index.js:7:31)
    at new TypeScript (/path/to/my/project/node_modules/broccoli-typescript-compiler/dist/plugin.js:324:27)
    at typescript (/path/to/my/project/node_modules/broccoli-typescript-compiler/index.js:28:10)
    at TypeScriptPreprocessor.toTree (/path/to/my/project/node_modules/ember-cli-typescript/lib/typescript-preprocessor.js:96:7)
    at /path/to/my/project/node_modules/ember-cli-preprocess-registry/preprocessors.js:180:26
    at Array.forEach (native)
    at processPlugins (/path/to/my/project/node_modules/ember-cli-preprocess-registry/preprocessors.js:178:11)
    at module.exports.preprocessJs (/path/to/my/project/node_modules/ember-cli-preprocess-registry/preprocessors.js:171:10)
    at EmberApp.appAndDependencies (/path/to/my/project/node_modules/ember-cli/lib/broccoli/ember-app.js:1110:27)
    at EmberApp.javascript (/path/to/my/project/node_modules/ember-cli/lib/broccoli/ember-app.js:1225:30)
    at EmberApp.toArray (/path/to/my/project/node_modules/ember-cli/lib/broccoli/ember-app.js:1662:12)
    at EmberApp.toTree (/path/to/my/project/node_modules/ember-cli/lib/broccoli/ember-app.js:1684:38)
    at module.exports (/path/to/my/project/ember-cli-build.js:29:13)
    at Builder.setupBroccoliBuilder (/path/to/my/project/node_modules/ember-cli/lib/models/builder.js:56:19)
    at new Builder (/path/to/my/project/node_modules/ember-cli/lib/models/builder.js:30:10)
    at ServeTask.run (/path/to/my/project/node_modules/ember-cli/lib/tasks/serve.js:24:55)
```

 I'm trying to use typescript, however this pops up now and then. Here is what I do to solve this:

1) `rm -rf tmp/`
2) `rm -rf node_modules/ && yarn install`
3) `yarn remove ember-cli-typescript` `ember install ember-cli-typescript`

If it pops up too often, I cancel on typescript and keep it deinstalled (however, yes, I'm really trying to use it)

### LINKED ISSUE #43: "Cannot set property 'noEmit' of undefined"
I get this error when trying to use e-c-ts in a fresh addon. Maybe I'm doing something wrong. Here's what I did:
1. Created my addon: `ember addon my-addon --yarn`
1. Installed e-c-ts: `ember install ember-cli-typescript`
1. Moved the e-c-ts dependency from `devDependencies` into `dependencies` as documented in the README
1. `yarn link`ed the dependency into my app and tried to run it

It seems that it's trying to load the tsconfig from the root of my app instead of from the addon
```
Cannot set property 'noEmit' of undefined
TypeError: Cannot set property 'noEmit' of undefined
    at TypeScriptPreprocessor.toTree (/Users/dwickern/code/my-addon/node_modules/ember-cli-typescript/lib/typescript-preprocessor.js:68:37)
    at /Users/dwickern/code/my-app/node_modules/ember-cli-preprocess-registry/preprocessors.js:180:26
    at Array.forEach (native)
    at processPlugins (/Users/dwickern/code/my-app/node_modules/ember-cli-preprocess-registry/preprocessors.js:178:11)
    at Function.module.exports.preprocessJs (/Users/dwickern/code/my-app/node_modules/ember-cli-preprocess-registry/preprocessors.js:171:10)
    at Class.preprocessJs (/Users/dwickern/code/my-app/node_modules/ember-cli/lib/models/addon.js:1085:25)
    at Class.processedAddonJsFiles (/Users/dwickern/code/my-app/node_modules/ember-cli/lib/models/addon.js:1099:33)
    at Class.compileAddon (/Users/dwickern/code/my-app/node_modules/ember-cli/lib/models/addon.js:1007:24)
    at Class.treeForAddon (/Users/dwickern/code/my-app/node_modules/ember-cli/lib/models/addon.js:661:26)
    at Class._treeFor (/Users/dwickern/code/my-app/node_modules/ember-cli/lib/models/addon.js:517:33)
```

### PR DESCRIPTION
**Pull request by chriskrycho**
This cleans up quite a few of the quirks I've seen personally—and a bunch of our open bugs!

- fixes #7
- fixes #23
- fixes #38
- fixes #43

### CODE REVIEW

**Comment by chriskrycho** [comment]
Cc. also @krisselden

**Comment by chriskrycho** [comment]
To be explicit about the problem: it's failing to resolve the TypeScript modules. 🤦‍♂️

**Comment by mfeckie** [comment]
The main difference I see just looking at this is that you're using a different version of the broccoli-typescript-compiler.  Other than that, I'm not seeing anything obvious.

File: lib/typescript-preprocessor.js [hunk 1]
```diff
@@ -1,102 +1,58 @@
...
-  }
 
-  toTree(inputNode, inputPath, outputPath) {
-    const tsconfig = readConfig(path.join(".", "tsconfig.json"));
+    const tsconfig = readConfig(path.join('.', 'tsconfig.json'))
 
     // The `include` setting is meant for the IDE integration; broccoli manages
     // manages its own input files.
-    tsconfig.include = ["**/*.ts"];
+    tsconfig.include = ['**/*.ts'];
```
Comment (chriskrycho) [inline lib/typescript-preprocessor.js]: Just needed to re-add this.

**Comment by chriskrycho** [comment]
Let's coordinate a release and announcement tomorrow, pending hearing from @krisselden that a full release is imminent.

**Comment by chriskrycho** [comment]
Two bugs I noted today in testing:

1. It does not properly resolve a number of paths. I have a bunch of things in a directory, `types`, at the root of my app. These resolve correctly in the base tsconfig; they do *not* resolve correctly in this add-on (presumably because of some configuration issue between this and `broccoli-typescript-compiler`).
2. The build process unavoidably spews *everything* to the screen. Given the above, and the poor state of affairs of Ember typings, that means that every rebuild is currently spewing hundreds (or possibly thousands) of errors to the screen.

I'd quite like to resolve both of those before officially releasing this, as it is *not* a good look. 😢

**Comment by krisselden** [comment]
@chriskrycho I'm going to need more info than that, is there a project i can clone to try it out?

**Comment by chriskrycho** [comment]
There's a reproduction [here](https://github.com/chriskrycho/ember-cli-typescript-bugs). Clone, yarn, and build should be all you need to see it. Locally, this is my output (after stripping out the `...Building` bits):

```
$ ember build
'instrument' is imported from external module 'ember-data/-debug' but never used
Warning: ignoring input sourcemap for vendor/ember/ember.debug.js because ENOENT: no such file or directory, open '/Users/chris/dev/oss/some-ts-bits/tmp/source_map_concat-input_base_path-Vyc1Zg2c.tmp/vendor/ember/ember.debug.map'
error TS5055: Cannot write file '/users/chris/dev/oss/some-ts-bits/config/environment.js' because it would overwrite input file.
  Adding a tsconfig.json file will help organize projects that contain both TypeScript and JavaScript files. Learn more at https://aka.ms/tsconfig.
error TS5055: Cannot write file '/users/chris/dev/oss/some-ts-bits/tests/helpers/destroy-app.js' because it would overwrite input file.
  Adding a tsconfig.json file will help organize projects that contain both TypeScript and JavaScript files. Learn more at https://aka.ms/tsconfig.
error TS5055: Cannot write file '/users/chris/dev/oss/some-ts-bits/tests/helpers/module-for-acceptance.js' because it would overwrite input file.
  Adding a tsconfig.json file will help organize projects that contain both TypeScript and JavaScript files. Learn more at https://aka.ms/tsconfig.
error TS5055: Cannot write file '/users/chris/dev/oss/some-ts-bits/tests/helpers/start-app.js' because it would overwrite input file.
  Adding a tsconfig.json file will help organize projects that contain both TypeScript and JavaScript files. Learn more at https://aka.ms/tsconfig.
some-ts-bits/tests/acceptance/some-ts-stuff-test.ts(1,22): error TS2307: Cannot find module 'qunit'.
some-ts-bits/tests/acceptance/some-ts-stuff-test.ts(7,3): error TS2304: Cannot find name 'server'.
some-ts-bits/tests/acceptance/some-ts-stuff-test.ts(10,11): error TS2345: Argument of type '<T>() => void' is not assignable to parameter of type '<T>(...args: any[]) => T'.
  Type 'void' is not assignable to type 'T'.
Warning: ignoring input sourcemap for vendor/ember/ember-testing.js because ENOENT: no such file or directory, open '/Users/chris/dev/oss/some-ts-bits/tmp/source_map_concat-input_base_path-ZjAZKQN7.tmp/vendor/ember/ember-testing.map'
cleaning up...
Built project successfully. Stored in "dist/".
```

File: lib/typescript-preprocessor.js [hunk 2]
```diff
@@ -1,50 +1,20 @@
...
-const find = require('broccoli-stew').find;
-const Funnel = require("broccoli-funnel");
-const MergeTrees = require("broccoli-merge-trees");
+const Funnel = require('broccoli-funnel');
+const MergeTrees = require('broccoli-merge-trees');
 const ts = require('typescript');
 const tsc = require('broccoli-typescript-compiler').typescript;
-const UnwatchedDir = require('broccoli-source').UnwatchedDir;
 
 function readConfig(configFile) {
```
Comment (chriskrycho) [inline lib/typescript-preprocessor.js]: Now defunct entirely; delete.

**Comment by chriskrycho** [comment]
Aaaaaand, now the tests pass but when I install it in an add-on, it's not generating an TS. (╯°□°）╯︵ ┻━┻

**Comment by chriskrycho** [comment]
Steps to reproduce this all-of-a-sudden-not-working:

```sh
ember new ts-pls
cd ts-pls
yarn add https://github.com/emberwatch/ember-cli-typescript.git#latest-b-c-ts
ember init ember-cli-typescript
ember generate component some-component
mv app/components/some-component.js app/components/some-component.ts
# add some content to some-component.ts and some-component.hbs
ember serve
```

Then proceed to beat head against wall as you can see that TS files *are* being compiled but *are not* being included in final output.

@krisselden am I just an idiot? (Entirely possible.)

**Comment by chriskrycho** [comment]
Note that the changes I just pushed do not fix this, but they do seem sensible.

**Pull request merged**